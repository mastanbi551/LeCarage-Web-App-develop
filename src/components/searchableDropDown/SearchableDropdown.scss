@import "../../styles//variables.scss";


.searchable-dropdown {
    position: relative;
    .static{
        &::placeholder{
        color:$primaryBlack;
    }}
    .inputdropdown{
        position: absolute;
    background-color: #fff;
    border: 1px solid #b4b4b4;
    z-index: 3;
    border-radius: 0px;
    padding-left: 0px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
    
    color: $primaryBlack;
    width: 100%;
    li{
        padding:3px;
        list-style-type: none;
        
    }

    li:hover{
        background-color:#02a8dd;
        color:#fff ;
    }
}}