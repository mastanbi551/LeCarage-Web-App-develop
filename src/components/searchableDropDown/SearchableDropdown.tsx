import _, { drop } from 'lodash';
import React, { useState, useRef, useEffect } from 'react';
import './SearchableDropdown.scss';

interface Option {
  label: string;
  value: string;
}

interface SearchableDropdownProps {
  options: Option[];
  labelName: string;
  handleChange: (e: React.ChangeEvent<HTMLInputElement> | undefined) => void;
  handleSelectChange: (e: React.FormEvent<HTMLLIElement> | undefined, value: string, label: string) => void;
  handleBlur?: (e: React.FocusEvent<HTMLInputElement, Element>) => void;
  error?: string | string[] | undefined;
  touched?: boolean | undefined;
  // onValueSelect: (
  //   e: React.FormEvent<HTMLLIElement>,
  //   value: string,
  //   label: string
  // ) => void;
  name: string;
  isDisable?: boolean | undefined;
  defaultValue?: string
  onFocus?: React.FocusEventHandler<HTMLLIElement> | undefined
}

export const SearchableDropdown: React.FC<SearchableDropdownProps> = ({
  options,
  labelName,
  handleChange,
  handleBlur,
  error,
  touched,
  name,
  isDisable,
  // onValueSelect,
  handleSelectChange,
  defaultValue,
  onFocus
}) => {
  const [inputValue, setInputValue] = useState<string | undefined>('');
  const [dropdownOptions, setDropdownOptions] = useState<Option[]>([]);
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const [dropdownValue, setDropdownValue] = useState<string | undefined>('');

  useEffect(() => {
    setDropdownOptions(options);
    const initialValue = options && options.length > 0 ? _.find(options, { value: defaultValue })?.label : '';
    setInputValue(initialValue);
  }, [options]);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('click', handleClickOutside);

    return () => {
      document.removeEventListener('click', handleClickOutside);
    };
  }, [dropdownRef]);

  useEffect(() => {
    if(defaultValue !== ''){
      setDropdownValue(defaultValue);
      //setInputValue(defaultValue);
    }
  }, [defaultValue]);


  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setInputValue(value);
  
    // Check if the input value exactly matches any existing option's label
    const selectedOption = options.find((option) =>
      option.label.toLowerCase() === value.toLowerCase()
    );
  
    if (selectedOption) {
      // If a match is found, set it as the selected option and close the dropdown
      setDropdownOptions([selectedOption]);
      setIsOpen(false);
      handleItemClick(selectedOption.label);
      handleSelectChange(undefined, selectedOption.value, selectedOption.label);
    } else {
      // If no match is found, filter the dropdown options based on the input value
      const filteredOptions = options.filter((option) =>
        option.label.toLowerCase().includes(value.toLowerCase())
      );
      setDropdownOptions(filteredOptions);
      setIsOpen(true);
    }
  };
  

  const handleItemClick = (value: string) => {
    setInputValue(value);
    setIsOpen(false);
  };

  return (
    <div className="searchable-dropdown input-text-field" ref={dropdownRef}>
      <input
        className={name + ' static'}
        type="text"
        value={defaultValue === '' ? '' : inputValue}
        onChange={(e) => {
          handleInputChange(e);
          handleChange(e);
        }}
        disabled={isDisable ? true : false}
        placeholder="Create/Select"
        onClick={() => setIsOpen(true)}
        onBlur={handleBlur}
        onKeyDown= {(e) => setIsOpen(true)}
        defaultValue={defaultValue}
      />
      
      {isOpen && dropdownOptions.length > 0 && (
        <ul className="inputdropdown" style={{ maxHeight: '200px', overflowY: 'auto' }}>
          {dropdownOptions.map((option, index) => (
            <li
              key={index}
              value={defaultValue !== '' ? defaultValue : option.label}
              defaultValue={defaultValue}
              onClick={(e) => {
                handleSelectChange(e, option.value, option.label);
                handleItemClick(option.label);
                // onValueSelect(e, option.value, option.label);                
              }}
              onFocus={onFocus}
            >
              {option.label}
            </li>
          ))}
        </ul>
      )}
      {error && touched ? (
        <div className="validation-error-message text-align-left">{error}</div>
      ) : null}
    </div>
  );
};
