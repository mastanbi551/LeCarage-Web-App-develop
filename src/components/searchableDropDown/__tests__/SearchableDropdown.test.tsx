import React from 'react';
import { render } from '@testing-library/react';
import { SearchableDropdown } from '../SearchableDropdown';

describe('SearchableDropdown', () => {
  test('renders correctly', () => {
    const props = {
      options: [
        { label: 'Option 1', value: 'option1' },
        { label: 'Option 2', value: 'option2' },
        { label: 'Option 3', value: 'option3' },
      ],
      labelName: 'Dropdown',
      handleChange: jest.fn(),
      handleSelectChange: jest.fn(),
      handleBlur: jest.fn(),
      error: undefined,
      touched: undefined,
      name: 'dropdown',
      isDisable: false,
      defaultValue: 'option2',
      onFocus: jest.fn(),
    };

    const { container } = render(<SearchableDropdown {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
