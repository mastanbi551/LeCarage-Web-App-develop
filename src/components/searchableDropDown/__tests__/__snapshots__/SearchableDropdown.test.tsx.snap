// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchableDropdown renders correctly 1`] = `
<div
  class="searchable-dropdown input-text-field"
>
  <input
    class="dropdown static"
    placeholder="Create/Select"
    type="text"
    value="Option 2"
  />
</div>
`;
